### ISSUE TITLE
Validate Gorilla/Mux Dependency

### ISSUE THREAD

**Issue by fforootd** [zitadel/zitadel#4285]
## Issue

It looks like the [gorilla/mux ](https://github.com/gorilla/mux/issues/659) projects is looking for new maintainers.

We need to validate how we should proceed in regard to this since it could proof a security risk.

## Options

- Contribute to the project
- Look for an alternative

**Comment by fforootd** [comment]
@zitadel/engineers opinions for this?

**Comment by adlerhurst** [comment]
Found a discussion here: https://twitter.com/itsharsh2001/status/1558748817414971392?s=21&t=GETNABbt8zWMI20WldqaNQ

**Comment by adlerhurst** [comment]
Opinion to the options you've mentioned:

- Contribute to the project: They are looking for a new maintainer, we would have to maintain the project. As there are many routers out there it feels like an overkill.
- Look for an alternative: There are many great routers out there. Feels like a good solution.
- Fork it: I'm not sure if we want to maintain a router. It's out of our scope and there are already great alternatives.

**Comment by livio-a** [comment]
I'd prefer an alternative. I don't think we have specific requirements for the router. Our current interceptors should not be dependant on gorilla that much.

**Comment by fforootd** [comment]
@livio-a @adlerhurst since the gorilla repos are now officialy in archive mode we need to actively work on this.

CC @hifabienne

**Comment by adlerhurst** [comment]
## Decision drivers

- dynamic routing (e.g. `users/{id}`)
- crsf handling (currently `gorilla/csrf`)
- form parsing

## Activaly maintained packages

- [https://pkg.go.dev/github.com/gin-gonic/gin](gin) (used widely in golang)
- [https://pkg.go.dev/github.com/go-chi/chi/v5](chi) (no dependencies)
- [https://pkg.go.dev/github.com/labstack/echo/v4](echo)
- [https://github.com/valyala/fasthttp](fasthttp)

**Comment by fforootd** [comment]
Some more inputs to this https://mattermost.com/blog/choosing-a-go-framework-gin-vs-echo/

**Comment by fforootd** [comment]
@adlerhurst I think we have more requirments. 

- Middleware to attach actions
- Also we want to serve html and "assets"
- Cors is also a thing to consider 😁

**Comment by adlerhurst** [comment]
- should be possible to add middlewares using `http.Handler` or `http.HandleFunc`

**Comment by fforootd** [comment]
This also depends on some changes plannend in #4839

**Comment by adlerhurst** [comment]
I would go with chi

**Comment by adlerhurst** [comment]
chi has no dependencies, no custom implementation of `Context`, feels nearly as the native `http.Router` with some additions

@livio-a correct me if I'm wrong but I think most of our middleware implementations should work out of the box. List of provided middlewares: https://github.com/go-chi/chi#middlewares 

Here's the discussion of gitea: https://github.com/go-gitea/gitea/issues/7403

If we switch to openapi there is a generator for the big golang frameworks: https://github.com/deepmap/oapi-codegen

**Comment by hifabienne** [comment]
@adlerhurst @livio-a So the decision is done? Can you write a user story according to this. So we can estimate and plan it?

**Comment by muhlemmer** [comment]
I agree with the point @adlerhurst  is making. I worked with Gin before and it's biggest down-side was context handling. Chi seems to be clean from this perspective. As a plus, there is also the `go-chi/render` package which can take care of [form parsing](https://pkg.go.dev/github.com/go-chi/render#DecodeForm) in place of `gorilla/schema`.

Maybe we should try it for `oidc` first? It is smaller as zitadel and allows us to validate the idea before spending a lot of hours on the zitadel side.

I can create a user story there if everyone agrees?

**Comment by adlerhurst** [comment]
I like the idea @muhlemmer I think it's easier to gain some knowledge of the router there

**Comment by muhlemmer** [comment]
Lessons learned from the router migration at zitadel/oidc#322:

- Replace `router := mux.NewRouter()` with `router := chi.NewRouter()` and it will work for all simple `HandlerFunc` routes.
- Same with the `router.Use()` function for middleware registration.
- Specific method routes require slight refactoring:
```
- l.router.Path("/username").Methods("GET").HandlerFunc(l.loginHandler) // mux
+ l.router.Get("/username", l.loginHandler) // chi
```
- Path prefix / sub-routing is slightly different. 
```
- router.PathPrefix("/login/").Handler(http.StripPrefix("/login", l.router)) // mux
+ router.Mount("/login/", http.StripPrefix("/login", l.router)) // chi
```

```
- registerDeviceAuth(storage, router.PathPrefix("/device").Subrouter()) // mux
+ router.Route("/device", func(r chi.Router) { // chi
	registerDeviceAuth(storage, r)
})
```
- URL path parameters are supported
```
/protected/{claim}/{value}
requestedClaim := chi.URLParam(r, "claim")
requestedValue := chi.URLParam(r, "value")
```

- URL query parameters are **not** supported, instead you need to use the `r.Form` to retrieve values. There is also no key based routing.

**Comment by elinashoko** [comment]
Thanks for filing this issue. As part of an ongoing backlog cleanup, we're closing issues that have had no activity in over 12 months.

If this is still relevant, let us know in a comment (or reopen the issue) and we'll take another look.

This issue will close in 30 days if there's no further activity.
